### PR TITLE
Fix #209

### LINKED ISSUE #209: Slave not feeding data to Master as smokeping_cgi mishandle the path to cache file which to lock.
`# less /usr/local/nginx/logs/error.log`
```
[Tue Dec 17 22:41:47 2019] smokeping.cgi [client 127.0.0.1]: Could not lock /var/lib/smokeping/__cgi//Asia/DMIT-hk.SLAVE_NAME.slave_cache (No such file or directory). Trying again 1 more times.
[Tue Dec 17 22:41:48 2019] smokeping.cgi [client 127.0.0.1]: Could not update /var/lib/smokeping/__cgi//Asia" while reading response header from upstream, client: 127.0.0.1, server: smokeping.domain.tld, request: "POST /smokeping.fcgi HTTP/1.1", upstream: "fastcgi://unix:/var/run/fcgiwrap.socket:", host: "smokeping.domain.tld"
2019/12/17 22:41:48 [error] 4380#4380: *90 FastCGI sent in stderr: "/DMIT-hk.SLAVE_NAME.slave_cache, giving up for now. at /usr/share/perl5/Smokeping/Master.pm line 156.
```
### Note that there is a redundent slash in the path `/var/lib/smokeping/__cgi//Asia` thus Master node failed to update rrd files while Slave keep feeding fresh data to Master without any failure.

That error message led me to the only reasonable suspetion that I had encounter a bug, not mis-configuration.

---
**nginx.conf**
```
http {
    include mime.types;
    error_log logs/error.log warn;
    server {
        listen ******;
        server_name smokeping.domain.tld;
        include fcgiwrap.conf;
        root /var/www/smokeping;
        index smokeping.fcgi;
        gzip off;
        location ~ \.fcgi$ {
            fastcgi_intercept_errors on;
            include fastcgi_params;
            fastcgi_param SCRIPT_FILENAME /var/www/smokeping/smokeping.fcgi;
            fastcgi_pass unix:/var/run/fcgiwrap.socket;
        }
        location ^~ /smokeping/images {
            alias /var/cache/smokeping/images;
            gzip off;
        }
    }
}
```

**/usr/share/perl5/Smokeping/Master.pm**
```
107 sub save_updates {
108     my $cfg = shift;
109     my $slave = shift;
110     my $updates = shift;
111     # name\ttime\tupdatestring
112     # name\ttime\tupdatestring
113     my %u;
114     for my $update (split /\n/, $updates){
115         my ($name, $time, $updatestring) = split /\t/, $update;
116         if ( ${name} =~ m{(^|/)\.\.($|/)} ){
117             warn "Skipping update for ${name}.${slave}.slave_cache since ".
118                  "you seem to try todo some directory magic here. Don't!";
119         } else {
120             push @{$u{$name}}, [$time,$updatestring];
121         }
122     }
123     for my $name (sort keys %u){
124         my $file = slavedatadir($cfg) ."/${name}.${slave}.slave_cache";
125         for (my $i = 2; $i >= 0; $i--){
126             my $fh;
127             if ( open ($fh, '+>>' , $file) and flock($fh, LOCK_EX) ){
128                 my $existing = [];
129                 if (! -e $file) { # the reader unlinked it from under us
130                     flock($fh, LOCK_UN);
131                     close $fh;
132                     next;
133                 }
134                 seek $fh, 0, 0;
135                 if ( -s _ ){
136                     my $in = eval { fd_retrieve $fh };
137                     if ($@) { #error
138                         warn "Loading $file: $@";
139                     } else {
140                         $existing = $in;
141                             };
142                 };
143                 map {
144                     push @{$existing}, [ $slave, $_->[0], $_->[1] ];
145                 } @{$u{$name}};
146                 nstore($existing, $file.$$);
147                rename $file.$$,$file;
148                 flock($fh, LOCK_UN);
149                 close $fh;
150                 last;
151             } elsif ($i > 0) {
152                 warn "Could not lock $file ($!). Trying again $i more times.\n";
153                 sleep rand(2);
154                 next;
155             }
156             warn "Could not update $file, giving up for now.";
157             close $fh;
158         }
159     }
160 };
```
---
**OS**: Ubuntu 19.10
**perl**: v5.28.1
**smokeping**:
```
$ smokeping --version
$ 2.007003
```
**nginx**: 
```
nginx version: nginx/1.17.6
built by gcc 7.4.0 (Ubuntu 7.4.0-1ubuntu1~18.04.1)
built with OpenSSL 1.1.1  11 Sep 2018 (running with OpenSSL 1.1.1c  28 May 2019)
TLS SNI support enabled
configure arguments: --user=www-data --group=www-data --with-threads --with-file-aio --with-http_ssl_module --with-http_v2_module --with-http_realip_module --with-http_image_filter_module=dynamic --with-http_geoip_module=dynamic --with-http_sub_module --without-mail_pop3_module --without-mail_imap_module --without-mail_smtp_module --with-stream --with-stream_ssl_module --with-stream_realip_module --with-stream_geoip_module --with-stream_ssl_preread_module --with-compat --with-pcre=/etc/icn/pcre-8.43/ --with-pcre-jit
```

### PR DESCRIPTION
**Pull request by IceCodeNew**
This fix has been confirmed by @rborgmaster.
Refer: https://github.com/oetiker/SmokePing/issues/209#issuecomment-679252929

### CODE REVIEW

File: lib/Smokeping/Master.pm [hunk 1]
```diff
@@ -121,7 +123,7 @@ sub save_updates {
         }
     }
     for my $name (sort keys %u){
-        my $file = slavedatadir($cfg) ."/${name}.${slave}.slave_cache";
+        my $file = slavedatadir($cfg) ."${name}.${slave}.slave_cache";
```
Comment (oetiker) [inline lib/Smokeping/Master.pm:126]:
```suggestion
        my $file = slavedatadir($cfg) ."/${name}.${slave}.slave_cache";
```

File: lib/Smokeping/Master.pm [hunk 2]
```diff
@@ -100,8 +100,10 @@ if 'dyndir' is not present.
 
 sub slavedatadir ($) {
     my $cfg = shift;
-    return $cfg->{General}{dyndir}  ||
+    my $dir = $cfg->{General}{dyndir}  ||
            $cfg->{General}{datadir};
+    $dir =~ s{/*$}{/};
```
Comment (oetiker) [inline lib/Smokeping/Master.pm:105]:
```suggestion
    $dir =~ s{/*$}{};
```

**Comment by IceCodeNew** [comment]
May I rebase now? Seems you have done your work on it.

**Comment by IceCodeNew** [comment]
Thanks for your work! @oetiker 
May you merge this PR into `master` branch?

**Comment by oetiker** [comment]
I will if you can confirm that it solves your problem

**Review by oetiker — APPROVED** [review]

**Pull request merged**